// @genkit-ai/vertexai/lib/anthropic.mjs
import {
  __async,
  __forAwait,
  __spreadProps,
  __spreadValues
} from "./chunk-WFI2LP4G.mjs";
import { AnthropicVertex } from "@anthropic-ai/vertex-sdk";
import {
  GenerationCommonConfigSchema,
  defineModel,
  getBasicUsageStats,
  modelRef
} from "@genkit-ai/ai/model";
import { GENKIT_CLIENT_HEADER } from "@genkit-ai/core";
import z from "zod";
const AnthropicConfigSchema = GenerationCommonConfigSchema.extend({
  location: z.string().optional()
});
const claude35Sonnet = modelRef({
  name: "vertexai/claude-3-5-sonnet",
  info: {
    label: "Vertex AI Model Garden - Claude 3.5 Sonnet",
    versions: ["claude-3-5-sonnet@20240620"],
    supports: {
      multiturn: true,
      media: true,
      tools: true,
      systemRole: true,
      output: ["text"]
    }
  },
  configSchema: AnthropicConfigSchema
});
const claude3Sonnet = modelRef({
  name: "vertexai/claude-3-sonnet",
  info: {
    label: "Vertex AI Model Garden - Claude 3 Sonnet",
    versions: ["claude-3-sonnet@20240229"],
    supports: {
      multiturn: true,
      media: true,
      tools: true,
      systemRole: true,
      output: ["text"]
    }
  },
  configSchema: AnthropicConfigSchema
});
const claude3Haiku = modelRef({
  name: "vertexai/claude-3-haiku",
  info: {
    label: "Vertex AI Model Garden - Claude 3 Haiku",
    versions: ["claude-3-haiku@20240307"],
    supports: {
      multiturn: true,
      media: true,
      tools: true,
      systemRole: true,
      output: ["text"]
    }
  },
  configSchema: AnthropicConfigSchema
});
const claude3Opus = modelRef({
  name: "vertexai/claude-3-opus",
  info: {
    label: "Vertex AI Model Garden - Claude 3 Opus",
    versions: ["claude-3-opus@20240229"],
    supports: {
      multiturn: true,
      media: true,
      tools: true,
      systemRole: true,
      output: ["text"]
    }
  },
  configSchema: AnthropicConfigSchema
});
const SUPPORTED_ANTHROPIC_MODELS = {
  "claude-3-5-sonnet": claude35Sonnet,
  "claude-3-sonnet": claude3Sonnet,
  "claude-3-opus": claude3Opus,
  "claude-3-haiku": claude3Haiku
};
function toAnthropicRequest(model, input) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k;
  let system = void 0;
  const messages = [];
  for (const msg of input.messages) {
    if (msg.role === "system") {
      system = msg.content.map((c) => {
        if (!c.text) {
          throw new Error(
            "Only text context is supported for system messages."
          );
        }
        return c.text;
      }).join();
    } else if (msg.content[msg.content.length - 1].toolResponse) {
      messages.push({
        role: "user",
        content: toAnthropicContent(msg.content)
      });
    } else {
      messages.push({
        role: toAnthropicRole(msg.role),
        content: toAnthropicContent(msg.content)
      });
    }
  }
  const request = {
    model,
    messages,
    // https://docs.anthropic.com/claude/docs/models-overview#model-comparison
    max_tokens: (_b = (_a = input.config) == null ? void 0 : _a.maxOutputTokens) != null ? _b : 4096
  };
  if (system) {
    request["system"] = system;
  }
  if (input.tools) {
    request.tools = (_c = input.tools) == null ? void 0 : _c.map((tool) => {
      return {
        name: tool.name,
        description: tool.description,
        input_schema: tool.inputSchema
      };
    });
  }
  if ((_d = input.config) == null ? void 0 : _d.stopSequences) {
    request.stop_sequences = (_e = input.config) == null ? void 0 : _e.stopSequences;
  }
  if ((_f = input.config) == null ? void 0 : _f.temperature) {
    request.temperature = (_g = input.config) == null ? void 0 : _g.temperature;
  }
  if ((_h = input.config) == null ? void 0 : _h.topK) {
    request.top_k = (_i = input.config) == null ? void 0 : _i.topK;
  }
  if ((_j = input.config) == null ? void 0 : _j.topP) {
    request.top_p = (_k = input.config) == null ? void 0 : _k.topP;
  }
  return request;
}
function toAnthropicContent(content) {
  return content.map((p) => {
    if (p.text) {
      return {
        type: "text",
        text: p.text
      };
    }
    if (p.media) {
      let b64Data = p.media.url;
      if (b64Data.startsWith("data:")) {
        b64Data = b64Data.substring(b64Data.indexOf(",") + 1);
      }
      return {
        type: "image",
        source: {
          type: "base64",
          data: b64Data,
          media_type: p.media.contentType
        }
      };
    }
    if (p.toolRequest) {
      return toAnthropicToolRequest(p.toolRequest);
    }
    if (p.toolResponse) {
      return toAnthropicToolResponse(p);
    }
    throw new Error(`Unsupported content type: ${JSON.stringify(p)}`);
  });
}
function toAnthropicRole(role) {
  if (role === "model") {
    return "assistant";
  }
  if (role === "user") {
    return "user";
  }
  if (role === "tool") {
    return "assistant";
  }
  throw new Error(`Unsupported role type ${role}`);
}
function fromAnthropicTextPart(part) {
  return {
    text: part.text
  };
}
function fromAnthropicToolCallPart(part) {
  return {
    toolRequest: {
      name: part.name,
      input: part.input,
      ref: part.id
    }
  };
}
function fromAnthropicPart(part) {
  if (part.type === "text")
    return fromAnthropicTextPart(part);
  if (part.type === "tool_use")
    return fromAnthropicToolCallPart(part);
  throw new Error(
    "Part type is unsupported/corrupted. Either data is missing or type cannot be inferred from type."
  );
}
function fromAnthropicCandidate(candidate) {
  const parts = candidate.content;
  const genkitCandidate = {
    index: 0,
    message: {
      role: "model",
      content: parts.map(fromAnthropicPart)
    },
    finishReason: toGenkitFinishReason(
      candidate.stop_reason
    ),
    custom: {
      id: candidate.id,
      model: candidate.model,
      type: candidate.type
    }
  };
  return genkitCandidate;
}
function fromAnthropicResponse(input, response) {
  const candidates = [fromAnthropicCandidate(response)];
  return {
    candidates,
    usage: __spreadProps(__spreadValues({}, getBasicUsageStats(input.messages, candidates)), {
      inputTokens: response.usage.input_tokens,
      outputTokens: response.usage.output_tokens
    })
  };
}
function toGenkitFinishReason(reason) {
  switch (reason) {
    case "end_turn":
      return "stop";
    case "max_tokens":
      return "length";
    case "stop_sequence":
      return "stop";
    case "tool_use":
      return "stop";
    case null:
      return "unknown";
    default:
      return "other";
  }
}
function toAnthropicToolRequest(tool) {
  if (!tool.name) {
    throw new Error("Tool name is required");
  }
  if (!/^[a-zA-Z0-9_-]{1,64}$/.test(tool.name)) {
    throw new Error(
      `Tool name ${tool.name} contains invalid characters. 
      Only letters, numbers, and underscores are allowed, 
      and the name must be between 1 and 64 characters long.`
    );
  }
  const declaration = {
    type: "tool_use",
    id: tool.ref,
    name: tool.name,
    input: tool.input
  };
  return declaration;
}
function toAnthropicToolResponse(part) {
  var _a;
  if (!((_a = part.toolResponse) == null ? void 0 : _a.ref)) {
    throw new Error("Tool response reference is required");
  }
  if (!part.toolResponse.output) {
    throw new Error("Tool response output is required");
  }
  return {
    type: "tool_result",
    tool_use_id: part.toolResponse.ref,
    content: JSON.stringify(part.toolResponse.output)
  };
}
function anthropicModel(modelName, projectId, region) {
  var _a, _b, _c;
  const clients = {};
  const clientFactory = (region2) => {
    if (!clients[region2]) {
      clients[region2] = new AnthropicVertex({
        region: region2,
        projectId,
        defaultHeaders: {
          "X-Goog-Api-Client": GENKIT_CLIENT_HEADER
        }
      });
    }
    return clients[region2];
  };
  const model = SUPPORTED_ANTHROPIC_MODELS[modelName];
  if (!model) {
    throw new Error(`unsupported Anthropic model name ${modelName}`);
  }
  return defineModel(
    {
      name: model.name,
      label: (_a = model.info) == null ? void 0 : _a.label,
      configSchema: AnthropicConfigSchema,
      supports: (_b = model.info) == null ? void 0 : _b.supports,
      versions: (_c = model.info) == null ? void 0 : _c.versions
    },
    (input, streamingCallback) => __async(this, null, function* () {
      var _a2, _b2, _c2, _d, _e;
      const client = clientFactory(((_a2 = input.config) == null ? void 0 : _a2.location) || region);
      if (!streamingCallback) {
        const response = yield client.messages.create(__spreadProps(__spreadValues({}, toAnthropicRequest((_c2 = (_b2 = input.config) == null ? void 0 : _b2.version) != null ? _c2 : modelName, input)), {
          stream: false
        }));
        return fromAnthropicResponse(input, response);
      } else {
        const stream = yield client.messages.stream(
          toAnthropicRequest((_e = (_d = input.config) == null ? void 0 : _d.version) != null ? _e : modelName, input)
        );
        try {
          for (var iter = __forAwait(stream), more, temp, error; more = !(temp = yield iter.next()).done; more = false) {
            const event = temp.value;
            if (event.type === "content_block_delta") {
              streamingCallback({
                index: 0,
                content: [
                  {
                    text: event.delta.text
                  }
                ]
              });
            }
          }
        } catch (temp) {
          error = [temp];
        } finally {
          try {
            more && (temp = iter.return) && (yield temp.call(iter));
          } finally {
            if (error)
              throw error[0];
          }
        }
        return fromAnthropicResponse(input, yield stream.finalMessage());
      }
    })
  );
}
export {
  AnthropicConfigSchema,
  SUPPORTED_ANTHROPIC_MODELS,
  anthropicModel,
  claude35Sonnet,
  claude3Haiku,
  claude3Opus,
  claude3Sonnet,
  fromAnthropicResponse,
  toAnthropicRequest
};
