// @genkit-ai/vertexai/lib/vector-search/indexers.mjs
import {
  __async
} from "../chunk-WFI2LP4G.mjs";
import { embedMany } from "@genkit-ai/ai/embedder";
import {
  defineIndexer,
  indexerRef
} from "@genkit-ai/ai/retriever";
import {
  Datapoint,
  VertexAIVectorIndexerOptionsSchema
} from "./types";
import { upsertDatapoints } from "./upsert_datapoints";
const vertexAiIndexerRef = (params) => {
  var _a;
  return indexerRef({
    name: `vertexai/${params.indexId}`,
    info: {
      label: (_a = params.displayName) != null ? _a : `Vertex AI - ${params.indexId}`
    },
    configSchema: VertexAIVectorIndexerOptionsSchema.optional()
  });
};
function vertexAiIndexers(params) {
  var _a;
  const vectorSearchOptions = params.pluginOptions.vectorSearchOptions;
  const defaultEmbedder = params.defaultEmbedder;
  const indexers = [];
  if (!vectorSearchOptions || vectorSearchOptions.length === 0) {
    return indexers;
  }
  for (const vectorSearchOption of vectorSearchOptions) {
    const { documentIndexer, indexId } = vectorSearchOption;
    const embedder = (_a = vectorSearchOption.embedder) != null ? _a : defaultEmbedder;
    const embedderOptions = vectorSearchOption.embedderOptions;
    const indexer = defineIndexer(
      {
        name: `vertexai/${indexId}`,
        configSchema: VertexAIVectorIndexerOptionsSchema.optional()
      },
      (docs, options) => __async(this, null, function* () {
        let docIds = [];
        try {
          docIds = yield documentIndexer(docs, options);
        } catch (error) {
          throw new Error(
            `Error storing your document content/metadata: ${error}`
          );
        }
        const embeddings = yield embedMany({
          embedder,
          content: docs,
          options: embedderOptions
        });
        const datapoints = embeddings.map(
          ({ embedding }, i) => new Datapoint({
            datapointId: docIds[i],
            featureVector: embedding
          })
        );
        try {
          yield upsertDatapoints({
            datapoints,
            authClient: params.authClient,
            projectId: params.pluginOptions.projectId,
            location: params.pluginOptions.location,
            indexId
          });
        } catch (error) {
          throw error;
        }
      })
    );
    indexers.push(indexer);
  }
  return indexers;
}
export {
  vertexAiIndexerRef,
  vertexAiIndexers
};
