// @genkit-ai/vertexai/lib/vector-search/index.mjs
import "../chunk-WFI2LP4G.mjs";
import {
  getBigQueryDocumentIndexer,
  getBigQueryDocumentRetriever
} from "./bigquery";
import {
  getFirestoreDocumentIndexer,
  getFirestoreDocumentRetriever
} from "./firestore";
import { vertexAiIndexerRef, vertexAiIndexers } from "./indexers";
import { vertexAiRetrieverRef, vertexAiRetrievers } from "./retrievers";
import {
  DocumentIndexer,
  DocumentRetriever,
  Neighbor,
  VectorSearchOptions,
  VertexAIVectorIndexerOptions,
  VertexAIVectorIndexerOptionsSchema,
  VertexAIVectorRetrieverOptions,
  VertexAIVectorRetrieverOptionsSchema
} from "./types";
export {
  DocumentIndexer,
  DocumentRetriever,
  Neighbor,
  VectorSearchOptions,
  VertexAIVectorIndexerOptions,
  VertexAIVectorIndexerOptionsSchema,
  VertexAIVectorRetrieverOptions,
  VertexAIVectorRetrieverOptionsSchema,
  getBigQueryDocumentIndexer,
  getBigQueryDocumentRetriever,
  getFirestoreDocumentIndexer,
  getFirestoreDocumentRetriever,
  vertexAiIndexerRef,
  vertexAiIndexers,
  vertexAiRetrieverRef,
  vertexAiRetrievers
};
