// @genkit-ai/vertexai/lib/vector-search/query_public_endpoint.mjs
import {
  __async
} from "../chunk-WFI2LP4G.mjs";
import { logger } from "@genkit-ai/core/logging";
function queryPublicEndpoint(params) {
  return __async(this, null, function* () {
    const {
      featureVector,
      neighborCount,
      accessToken,
      indexEndpointId,
      publicDomainName,
      projectNumber,
      deployedIndexId,
      location
    } = params;
    const url = new URL(
      `https://${publicDomainName}/v1/projects/${projectNumber}/locations/${location}/indexEndpoints/${indexEndpointId}:findNeighbors`
    );
    const requestBody = {
      deployed_index_id: deployedIndexId,
      queries: [
        {
          datapoint: {
            datapoint_id: "0",
            feature_vector: featureVector
          },
          neighbor_count: neighborCount
        }
      ]
    };
    const response = yield fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${accessToken}`
      },
      body: JSON.stringify(requestBody)
    });
    if (!response.ok) {
      logger.error("Error querying index: ", response.statusText);
    }
    return response.json();
  });
}
export {
  queryPublicEndpoint
};
