// @genkit-ai/core/lib/telemetry.mjs
import "./chunk-XEFTB2OF.mjs";
import { ValueType } from "@opentelemetry/api";
import { GENKIT_VERSION } from "./index.js";
import {
  internalMetricNamespaceWrap,
  MetricCounter,
  MetricHistogram
} from "./metrics.js";
import {
  spanMetadataAls,
  traceMetadataAls
} from "./tracing/instrumentation.js";
const _N = internalMetricNamespaceWrap.bind(null, "action");
const actionCounter = new MetricCounter(_N("requests"), {
  description: "Counts calls to genkit actions.",
  valueType: ValueType.INT
});
const actionLatencies = new MetricHistogram(_N("latency"), {
  description: "Latencies when calling Genkit actions.",
  valueType: ValueType.DOUBLE,
  unit: "ms"
});
function writeActionSuccess(actionName, latencyMs) {
  var _a, _b, _c, _d;
  const dimensions = {
    name: actionName,
    flowName: (_b = (_a = traceMetadataAls) == null ? void 0 : _a.getStore()) == null ? void 0 : _b.flowName,
    path: (_d = (_c = spanMetadataAls) == null ? void 0 : _c.getStore()) == null ? void 0 : _d.path,
    status: "success",
    source: "ts",
    sourceVersion: GENKIT_VERSION
  };
  actionCounter.add(1, dimensions);
  actionLatencies.record(latencyMs, dimensions);
}
function writeActionFailure(actionName, latencyMs, err) {
  var _a, _b, _c, _d;
  const dimensions = {
    name: actionName,
    flowName: (_b = (_a = traceMetadataAls) == null ? void 0 : _a.getStore()) == null ? void 0 : _b.flowName,
    path: (_d = (_c = spanMetadataAls) == null ? void 0 : _c.getStore()) == null ? void 0 : _d.path,
    source: "ts",
    sourceVersion: GENKIT_VERSION,
    status: "failure",
    error: err == null ? void 0 : err.name
  };
  actionCounter.add(1, dimensions);
  actionLatencies.record(latencyMs, dimensions);
}
export {
  writeActionFailure,
  writeActionSuccess
};
